accessor/range.hpp

```diff
@@ -69,7 +69,8 @@ class range {
      *
      * @param params  parameters forwarded to Accessor constructor.
      */
-    template <typename... AccessorParams>
+    template <typename... AccessorParams,
+              std::enable_if_t<(sizeof...(AccessorParams) > 1), bool> = true>
```

Comment:
This change would make the `include/ginkgo/core/base/range.hpp` diverge ? Maybe we should either note this somewhere and in 2.0 fix it, or adapt the `include/ginkgo/core/base/range.hpp` as well.

Reply:
what do you mean diverge here? the usage of constructor keeps the same due to the accessor always require more than 1 args

Reply:
I believe this is a duplication of the `include/ginkgo/core/base/range.hpp`, as it was needed to be separate for the Accessor. I meant now the typically the same class (range) has two different versions and we should probably unify them. Also I think you are changing one of the copy constructors below ?

Reply:
Yes, it uses enable_if_t to avoid compiler use this constructor as copy constuctor. By doing so, compiler can auto generate the copy/move constructor, so we do not need to manually point copy constructor to default behavior.
I can also change the part in `include/ginkgo/core/base/range.hpp`

Reply:
Yeah, I guess we need to be sure that it does not break interface, I would also maybe ping @thoasm for his comments here.

Reply:
I do not think it break any interface. Do we have two range class for different purpose?

Reply:
Divergence between the two should not matter much. The accessor in `accessor/` is not considered a public interface, so interface breaking changes in there are fine.
However, I dislike the current change as it prevents any accessor of taking a single argument. I will shortly post a solution to that (basically making sure that the argument given is not a range, but forwarding everything else to the accessor constructor).

Reply:
I see. I can also do that if you want

Reply:
Why exactly is it actually necessary that the compiler generates the copy constructor instead of setting it as `= default`?
We could definitively write the condition more specific, so only single range objects are rejected, but to me this feels like a lot of effort for simply avoiding writing `= default` (I would reject the current version, however, because you forbid an accessor with only a single argument).

Reply:
for dpcpp usage, I need to use two version copy constructor
1. declare and impl them in the class
```
range(const range &other) = default;
range(range &other) = default;
```
linux is okay if declare and define them in the class
but MSVC is not happy. please see [ref1](https://developercommunity.visualstudio.com/t/false-error-c2580-msvc-doesnt-allow-a-constructor-1/1271214), [ref2](https://developercommunity.visualstudio.com/t/error-c2580-multiple-versions-of-a-defaulted-speci/468431) and some stackoverflow
MSVC only fixes it in /c++latest

2. declare in the class but impl out of the class
```
class range{...};
template <class Accessor>
range<Accessor>(const range &other) = default;
template <class Accessor>
range<Accessor>(range &other) = default;
```
the workaround is mentioned in MSVC such that use two default on copy constructor.
However, dpcpp can not recognize the copy constructor is default behavior, so dpcpp will gives an error about the class is not trivial ...

3. make all copy constructor generated by compiler
the current one is implemented by this way. Avoid the templated constructor is seen as copy constructor such that compiler can automatically handle it.
MSVC and DPCPP are happy with this.

Reply:
Can you give a bit of background why we need a copy constructor and a mutable lvalue ref constructor?

Reply:
disabling Outer(const Outer&) 
```
error: no matching constructor for initialization of 'Outer'
       run([&]() { [=]() { func(a); }; });
                                ^
candidate constructor not viable: 1st argument ('const Outer') would lose const qualifier
```
disabling Outer(Outer&) 
```
note: in instantiation of function template specialization 'Outer::Outer<Outer &>' requested here
  run([&]() { [=]() { func(a); }; });
```
this is simple version for emulating the dpcpp submit the job.
As far as I remember, two default impl out of class is okay in C++ code, but dpcpp can not allow it.

Reply:
Can you post a longer snippet that contains the entire context? Why do you need to modify the element you are copying in a copy constructor?

Reply:
Right, we talked about something like that previously.
If I remember correctly, the problem is that you need both `const range &` and `range &` for them to be used, otherwise, for a `range &`, the forwarding constructor is taken. Additionally, for some reason, the visual studio compiler does not accept both copy constructors `range(const range &) = default` and `range(range &) = default` (apparently, implementing them instead of `=default` does not help as well).
Was that about right, @yhmtsai?
I would prefer to find an alternative solution and will try something tomorrow, otherwise, we can add the SFINAE work-around to only exclude single `range` parameter for this constructor.